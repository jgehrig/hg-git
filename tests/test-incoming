#!/bin/sh

# Fails for some reason, need to investigate
# "$TESTDIR/hghave" git || exit 80

# bail if the user does not have dulwich
python -c 'import dulwich, dulwich.repo' || exit 80

# This test only works on hg 1.7 and later
python -c 'from mercurial import util ; assert util.version() > "1.7"' || exit 80

# bail early if the user is already running git-daemon
echo hi | nc localhost 9418 2>/dev/null && exit 80

echo "[extensions]" >> $HGRCPATH
echo "hggit=$(echo $(dirname $(dirname $0)))/hggit" >> $HGRCPATH
echo 'hgext.graphlog =' >> $HGRCPATH
echo 'hgext.bookmarks =' >> $HGRCPATH

GIT_AUTHOR_NAME='test'; export GIT_AUTHOR_NAME
GIT_AUTHOR_EMAIL='test@example.org'; export GIT_AUTHOR_EMAIL
GIT_AUTHOR_DATE="2007-01-01 00:00:00 +0000"; export GIT_AUTHOR_DATE
GIT_COMMITTER_NAME="$GIT_AUTHOR_NAME"; export GIT_COMMITTER_NAME
GIT_COMMITTER_EMAIL="$GIT_AUTHOR_EMAIL"; export GIT_COMMITTER_EMAIL
GIT_COMMITTER_DATE="$GIT_AUTHOR_DATE"; export GIT_COMMITTER_DATE

count=10
commit()
{
    GIT_AUTHOR_DATE="2007-01-01 00:00:$count +0000"
    GIT_COMMITTER_DATE="$GIT_AUTHOR_DATE"
    git commit "$@" >/dev/null 2>/dev/null || echo "git commit error"
    count=`expr $count + 1`
}
hgcommit()
{
    HGDATE="2007-01-01 00:00:$count +0000"
    hg commit -d "$HGDATE" "$@" >/dev/null 2>/dev/null || echo "hg commit error"
    count=`expr $count + 1`
}

mkdir gitrepo
cd gitrepo
git init | python -c "import sys; print sys.stdin.read().replace('$(dirname $(pwd))/', '')"

echo alpha > alpha
git add alpha
commit -m "add alpha"

# dulwich does not presently support local git repos, workaround
cd ..
git daemon --base-path="$(pwd)"\
 --listen=localhost\
 --export-all\
 --pid-file="$DAEMON_PIDS" \
 --detach --reuseaddr \
 --enable=receive-pack

hg clone git://localhost/gitrepo hgrepo | grep -v '^updating'

cd hgrepo
hg incoming

cd ../gitrepo
echo beta > beta
git add beta
commit -m 'add beta'

cd ../hgrepo
hg incoming

cd ../gitrepo
git checkout -b b1 HEAD^
mkdir d
echo gamma > d/gamma
git add d/gamma
commit -m'add d/gamma'
git tag t1

echo gamma 2 >> d/gamma
git add d/gamma
commit -m'add d/gamma line 2'

cd ../hgrepo
hg incoming -p

echo % incoming -r
hg incoming -r master
hg incoming -r b1
hg incoming -r t1

echo % nothing incoming after pull
hg pull
hg incoming

echo 'done'
