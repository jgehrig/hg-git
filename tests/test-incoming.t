Load commonly used test logic
  $ . "$TESTDIR/testutil"

  $ git init gitrepo
  Initialized empty Git repository in $TESTTMP/gitrepo/.git/
  $ cd gitrepo
  $ echo alpha > alpha
  $ git add alpha
  $ fn_git_commit -m "add alpha"

  $ cd ..
  $ hg clone gitrepo hgrepo | grep -v '^updating'
  importing git objects into hg
  1 files updated, 0 files merged, 0 files removed, 0 files unresolved

  $ hg -R hgrepo incoming | grep -v 'no changes found' | grep -v 'bookmark:'
  comparing with $TESTTMP/gitrepo

  $ cd gitrepo
  $ echo beta > beta
  $ git add beta
  $ fn_git_commit -m 'add beta'
  $ cd ..

  $ hg -R hgrepo incoming | grep -v 'no changes found' | grep -v 'bookmark:'
  comparing with $TESTTMP/gitrepo
  changeset:   1:9497a4ee62e1
  user:        test <test@example.org>
  date:        Mon Jan 01 00:00:11 2007 +0000
  summary:     add beta
  

  $ cd gitrepo
  $ git checkout -b b1 HEAD^
  Switched to a new branch 'b1'
  $ mkdir d
  $ echo gamma > d/gamma
  $ git add d/gamma
  $ fn_git_commit -m'add d/gamma'
  $ git tag t1

  $ echo gamma 2 >> d/gamma
  $ git add d/gamma
  $ fn_git_commit -m'add d/gamma line 2'
  $ cd ../hgrepo
  $ hg incoming -p | grep -v 'no changes found' | grep -v 'bookmark:'
  comparing with $TESTTMP/gitrepo
  changeset:   1:9497a4ee62e1
  user:        test <test@example.org>
  date:        Mon Jan 01 00:00:11 2007 +0000
  summary:     add beta
  
  diff -r 3442585be8a6 -r 9497a4ee62e1 beta
  --- /dev/null	Thu Jan 01 00:00:00 1970 +0000
  +++ b/beta	Mon Jan 01 00:00:11 2007 +0000
  @@ -0,0 +1,1 @@
  +beta
  
  changeset:   2:9865e289be73
  tag:         t1
  parent:      0:3442585be8a6
  user:        test <test@example.org>
  date:        Mon Jan 01 00:00:12 2007 +0000
  summary:     add d/gamma
  
  diff -r 3442585be8a6 -r 9865e289be73 d/gamma
  --- /dev/null	Thu Jan 01 00:00:00 1970 +0000
  +++ b/d/gamma	Mon Jan 01 00:00:12 2007 +0000
  @@ -0,0 +1,1 @@
  +gamma
  
  changeset:   3:5202f48c20c9
  user:        test <test@example.org>
  date:        Mon Jan 01 00:00:13 2007 +0000
  summary:     add d/gamma line 2
  
  diff -r 9865e289be73 -r 5202f48c20c9 d/gamma
  --- a/d/gamma	Mon Jan 01 00:00:12 2007 +0000
  +++ b/d/gamma	Mon Jan 01 00:00:13 2007 +0000
  @@ -1,1 +1,2 @@
   gamma
  +gamma 2
  

incoming -r
  $ hg incoming -r master | grep -v 'no changes found' | grep -v 'bookmark:'
  comparing with $TESTTMP/gitrepo
  changeset:   1:9497a4ee62e1
  user:        test <test@example.org>
  date:        Mon Jan 01 00:00:11 2007 +0000
  summary:     add beta
  
  $ hg incoming -r b1 | grep -v 'no changes found' | grep -v 'bookmark:'
  comparing with $TESTTMP/gitrepo
  changeset:   1:9865e289be73
  tag:         t1
  user:        test <test@example.org>
  date:        Mon Jan 01 00:00:12 2007 +0000
  summary:     add d/gamma
  
  changeset:   2:5202f48c20c9
  user:        test <test@example.org>
  date:        Mon Jan 01 00:00:13 2007 +0000
  summary:     add d/gamma line 2
  
  $ hg incoming -r t1 | grep -v 'no changes found' | grep -v 'bookmark:'
  comparing with $TESTTMP/gitrepo
  changeset:   1:9865e289be73
  tag:         t1
  user:        test <test@example.org>
  date:        Mon Jan 01 00:00:12 2007 +0000
  summary:     add d/gamma
  

nothing incoming after pull
"adding remote bookmark" message was added in Mercurial 2.3
  $ hg pull | grep -v "adding remote bookmark"
  pulling from $TESTTMP/gitrepo
  importing git objects into hg
  (run 'hg heads' to see heads, 'hg merge' to merge)
  $ hg incoming | grep -v 'no changes found' | grep -v 'bookmark:'
  comparing with $TESTTMP/gitrepo
