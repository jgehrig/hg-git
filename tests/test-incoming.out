Initialized empty Git repository in gitrepo/.git/

Counting objects: 3, done.
Total 3 (delta 0), reused 0 (delta 0)
importing git objects into hg
1 files updated, 0 files merged, 0 files removed, 0 files unresolved
comparing with git://localhost/gitrepo
no changes found
comparing with git://localhost/gitrepo
Counting objects: 3, done.
Compressing objects:  50% (1/2)   Compressing objects: 100% (2/2)   Compressing objects: 100% (2/2), done.
Total 3 (delta 0), reused 0 (delta 0)
changeset:   1:9497a4ee62e1
bookmark:    master
user:        test <test@example.org>
date:        Mon Jan 01 00:00:11 2007 +0000
summary:     add beta

Switched to a new branch 'b1'
comparing with git://localhost/gitrepo
Counting objects: 8, done.
Compressing objects:  25% (1/4)   Compressing objects:  50% (2/4)   Compressing objects:  75% (3/4)   Compressing objects: 100% (4/4)   Compressing objects: 100% (4/4), done.
Total 8 (delta 0), reused 0 (delta 0)
changeset:   1:9497a4ee62e1
bookmark:    master
user:        test <test@example.org>
date:        Mon Jan 01 00:00:11 2007 +0000
summary:     add beta

diff -r 3442585be8a6 -r 9497a4ee62e1 beta
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/beta	Mon Jan 01 00:00:11 2007 +0000
@@ -0,0 +1,1 @@
+beta

changeset:   2:9865e289be73
tag:         t1
parent:      0:3442585be8a6
user:        test <test@example.org>
date:        Mon Jan 01 00:00:12 2007 +0000
summary:     add d/gamma

diff -r 3442585be8a6 -r 9865e289be73 d/gamma
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/d/gamma	Mon Jan 01 00:00:12 2007 +0000
@@ -0,0 +1,1 @@
+gamma

changeset:   3:5202f48c20c9
bookmark:    b1
user:        test <test@example.org>
date:        Mon Jan 01 00:00:13 2007 +0000
summary:     add d/gamma line 2

diff -r 9865e289be73 -r 5202f48c20c9 d/gamma
--- a/d/gamma	Mon Jan 01 00:00:12 2007 +0000
+++ b/d/gamma	Mon Jan 01 00:00:13 2007 +0000
@@ -1,1 +1,2 @@
 gamma
+gamma 2

% incoming -r
comparing with git://localhost/gitrepo
changeset:   1:9497a4ee62e1
bookmark:    master
user:        test <test@example.org>
date:        Mon Jan 01 00:00:11 2007 +0000
summary:     add beta

comparing with git://localhost/gitrepo
changeset:   1:9865e289be73
tag:         t1
user:        test <test@example.org>
date:        Mon Jan 01 00:00:12 2007 +0000
summary:     add d/gamma

changeset:   2:5202f48c20c9
bookmark:    b1
user:        test <test@example.org>
date:        Mon Jan 01 00:00:13 2007 +0000
summary:     add d/gamma line 2

comparing with git://localhost/gitrepo
changeset:   1:9865e289be73
tag:         t1
user:        test <test@example.org>
date:        Mon Jan 01 00:00:12 2007 +0000
summary:     add d/gamma

% nothing incoming after pull
pulling from git://localhost/gitrepo
importing git objects into hg
(run 'hg heads' to see heads, 'hg merge' to merge)
comparing with git://localhost/gitrepo
no changes found
done
